### DISCUSSION TITLE
Glide | Header Authentication, Setting Disk Cache and Persistent Cache

### DISCUSSION THREAD

**Discussion by PATRICKdallat** [dotnet/maui#18038] [Upgrading from Xamarin]
I am currently in the process of migrating a Xamarin project, alongside my team. Right now, we're using GlideX.Forms for Android Image Loading.

In our current setup, we've got Glide set up to write to a directory outside of the app's cache directory to steer clear of garbage collection issues. On top of that, we're adding authentication header values to our URI image requests.

I've been diving into the .NET MAUI codebase, but I've hit a couple of roadblocks. I'm hoping you can shed some light on these questions:

Is there a way to set the disk cache directory during app startup, either before or when Glide gets going?

Can we append authentication header values before the PlatformInterop process kicks in?

As a bonus, I'm curious about whether we have to keep the cache directory outside the app to ensure cache doesn't get wiped out by garbage collection. Or maybe there's a nifty way to flag cache as persistent within .NET MAUI?

@jonathanpeppers Your insights on these issues would be a huge help for our project. Thanks a ton for your time and expertise.

**Comment by jonathanpeppers** [comment]
The Glide implementation for MAUI is all in Java:

https://github.com/dotnet/maui/tree/main/src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/glide

@mattleibow @Redth is there any exposed way to access the underlying Glide APIs?

An example I implemented in glidex was this "random alpha" handler:

https://github.com/jonathanpeppers/glidex/blob/main/glidex.forms.sample/RandomAlphaHandler.cs

It was just a way to test devs had a way to configure Glide.

**Reply by markmccaigue** [reply]
It would be really great to know if such a mechanism exists.

One use case for us would be authentication. Ideally one could configure Glide to add a header to the request, but in the absence of such a configuration mechanism we might consider appending a token to the URL. This unfortunately also seems to be problematic as presumably new cache entries will be created each time the token rotates as the URLs will be changed. It seems this could also be mitigated if there were a configuration entry point, as a custom GlideUrl could be provided which could elect to exclude the token from the cache key.

In the absence of a way to configure one of these two elements, or some alternate approach, it's not clear that it's possible currently to load authenticated images on Android without running into the above cache issue?
